ethcore/src/client/client.rs

```diff
@@ -276,6 +278,25 @@ impl Client {
 		self.notify.write().push(Arc::downgrade(&target));
 	}
 
+	/// Set a closure to call when we want to restart the client
+	pub fn set_exit_handler<F>(&self, f: F) where F: Fn(bool, Option<String>) + 'static + Send {
+		*self.exit_handler.lock() = Some(Box::new(f));
+	}
+
+	/// Hypervisor should restart this client.
+	pub fn restart(&self) {
+		if let Some(ref h) = *self.exit_handler.lock() {
+			(*h)(true, None);
+		}
+	}
+
+	/// Hypervisor should kill this client.
+	pub fn exit(&self) {
```

Comment:
I don't see how it belongs in `Client`, whose role is just to manage and provide blockchain data. Hypervisor stuff being at a higher level would make it a lot simpler to manage different kinds of clients, parity-as-a-library, non-hypervised instances, etc.